### PR TITLE
Feat/make dfschema wrap schemaref

### LINKED ISSUE #4680: Make DfSchema wrap SchemaRef
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Currently DataFusion has two ways to represent a schema, `Schema` and `DFSchema`. The former is the representation used by arrow, and in most components. `DFSchema` appears to "enhance" an arrow schema with the notion of a qualifier.

I'm not entirely sure of the history of this split, but to the uninitiated the split is confusing and frustrating. It also results in a non-trivial amount of schema munging logic to convert to/from the relevant representations

**Describe the solution you'd like**

I would like to change `DfSchema` to be

```
struct DfSchema {
    schema: SchemaRef,
    fields: Vec<DfFieldMetadata>
}

struct DfFieldMetadata {
    qualifier: Option<String>,
}
```

We could then make `DfSchema` automatically deref to `SchemaRef`, or at the very least implement `AsRef<SchemaRef>`, avoiding a lot of code that ends up looking like

```
let schema: Schema = self.plan.schema().as_ref().into();
Arc::new(schema)
```

Components wishing to combine the information can easily zip the two together, we could even assist this by adding

```
struct DfField<'a> {
    field: &'a Field,
    metadata: &'a DfFieldMetadata
}

impl DfSchema {

    fn df_fields() -> impl Iterator<Item=DfField<'_>> + '_ {
        self.arrow_schema.fields().iter().zip(&self.fields).map(|(field, metadata)| DfField { field, metadata })
    }
}
```

**Describe alternatives you've considered**

We could not do this

**Additional context**

### PR DESCRIPTION
**Pull request by matthewmturner**
## Which issue does this PR close?

Closes #4680

### CODE REVIEW

**Comment by matthewmturner** [comment]
@alamb FYI adding for visibility. Still plenty of work to do but publishing in case theres any feedback along the way.

**Comment by matthewmturner** [comment]
I have come across several methods which have been deprecated for a while now (i.e. `Column.normalize_with_schema` (since v20), `DFSchema::new` (since v7), and `DFSchema.index_of` (v8)).  Is it safe to remove these now?

**Comment by matthewmturner** [comment]
As a status update this week has been slow as I've been on site for company hackathon. I expect to pick up again this weekend or early next week.

**Comment by matthewmturner** [comment]
@alamb would you be able to give this a quick overview to make sure its going in the right direction?

File: datafusion/common/src/dfschema.rs [hunk 1]
```diff
@@ -106,10 +107,11 @@ pub type DFSchemaRef = Arc<DFSchema>;
...
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub struct DFSchema {
-    /// Fields
-    fields: Vec<DFField>,
-    /// Additional metadata in form of key value pairs
-    metadata: HashMap<String, String>,
+    /// Inner Arrow schema reference.
+    inner: SchemaRef,
+    /// Optional qualifiers for each column in this schema. In the same order as
+    /// the `self.inner.fields()`
```
Comment (alamb) [inline datafusion/common/src/dfschema.rs:113]: 👍

**Review by alamb — COMMENTED** [review]
>  @alamb would you be able to give this a quick overview to make sure its going in the right direction?

I think it is is @matthewmturner  -- thank you 🙏 

Is there anything thing I can do to help (e.g. help port an API or debug tests?)

**Comment by matthewmturner** [comment]
@alamb great, thanks for checking it out.  all good for now.  i just got datafusion common to build after all these changes and now cleaning up a few failing tests.

**Comment by matthewmturner** [comment]
@alamb sorry this is taking a while, i havent had a significant amount of time to crank this out so im basically working on it bit by bit every morning i get the chance.  hopefully this isnt slowing anything down on your side.

**Comment by alamb** [comment]
> @alamb sorry this is taking a while, i havent had a significant amount of time to crank this out so im basically working on it bit by bit every morning i get the chance. hopefully this isnt slowing anything down on your side.

Thank you for keep plugging away. This is not blocking anything on my end (cc @comphead  who maybe is more actively working on the planning speed)

I think I would classify this PR as important but not urgent (much like splitting up functions into crates -- e.g. https://github.com/apache/arrow-datafusion/pull/9113). 

Thank you for continuing to push

**Comment by haohuaijin** [comment]
Hi @matthewmturner I've noticed that this PR has been open for quite a while and it seems like you've already done a lot of work on it. I'm really interested in this improve and would like to contribute if possible. If you're busy with other commitments or if you could use some help to finish this up, I'd be more than happy to assist. Of course, I want to respect your work, so please let me know how you would feel about this. Thanks for all the effort you've already put into this PR!

**Comment by matthewmturner** [comment]
@haohuaijin its very timely that you reached out on this I was actually about to post an update here that while I had planned to continue working on this when I had time I was going to be prioritizing some other work before this so I was open to others helping to push this forward during that time.  So I would be more than happy to have you assist getting this over the finish line :)

**Comment by haohuaijin** [comment]
@matthewmturner, thank you very much for your reply and openness to my participation in this PR. I'm excited for the opportunity to help move this project forward.
In the next work, I have two possible ways to proceed. One is that I push my updates to your repository, and then you can merge my PR into the current PR. The other is that I create a new PR based on your PR. I was wondering if you have any preference between these two methods? Or do you have any other suggestions?

**Comment by matthewmturner** [comment]
I dont want to hold you up if I am unable to merge your PR into mine for whatever reason so I think creating a new PR is probably the better approach here.  I look forward to seeing your work on this :)

**Comment by haohuaijin** [comment]
@matthewmturner, thank you very much for your advice and support. I will follow your suggestion and create a new PR based on your PR. I'll start working on it as soon as possible and let you know when it's complete. Thanks again for your help and support.

**Comment by haohuaijin** [comment]
@matthewmturner and @alamb, I've created a new PR (#9595) based on the original one. I welcome your review and feedback at any time. I will continue to work on #9595. If you have any questions or suggestions, please feel free to share them. I appreciate your guidance and support.

**Pull request closed without merging**